lib/stagger/tcp_client.js

```diff
@@ -0,0 +1,137 @@
+var events = require("events");
+var net = require("net");
+var util = require("util");
+
+var Parser = require('../protocol/parser');
+var encode = require('../protocol/encoder');
+
+// Manages a TCP connection to stagger, including reconnects
+// (De)serialises incoming/outgoing messages
+function TcpClient(tags, connectionOpts, timeout) {
+  events.EventEmitter.call(this);
+
+  this.timeout = timeout || 60000;
+  this.connected = false;
+
+  this.parser = new Parser(
+    function(method, body) {
+      switch (method) {
+        case 'pair:ping':
+          this.pong();
+          break;
+        case 'pair:pong':
+          break; // NOOP
+        default:
+          this.emit("message", method, body);
+          break;
+      }
+    }.bind(this),
+    function(err) {
+      console.log("Parse error: " + err);
+      this.socket.end(); // will trigger a close event, which will lead to a reconnect
+      this.handleShutdown();
+      this.parser.reset();
+    }.bind(this)
+  );
+
+  this.connect(tags, connectionOpts);
+}
+
+util.inherits(TcpClient, events.EventEmitter);
+var prototype = TcpClient.prototype;
+
+prototype.send = function(method, body) {
+  this.socket.write(encode(method, body));
+};
+
+// private
+
+prototype.connect = function(tags, connectionOpts) {
```

Comment:
Because I'm used to functional languages :P

I don't feel too strongly about it in this case, and want to stick to conventions for a particular language. So I'll make them object members if we think it makes more sense.